README.md
BlockTrail NodeJS SDK
=====================
This is the BlockTrail NodeJS SDK. This SDK contains methods for easily interacting with the BlockTrail API.
Below are examples to get you started. For additional examples, please see our official documentation
at https://www.blocktrail.com/api/docs/lang/nodejs

[![Latest Stable Version](https://badge.fury.io/js/blocktrail-sdk.svg)](https://www.npmjs.org/package/blocktrail-sdk)
[![Build Status](https://travis-ci.org/blocktrail/blocktrail-sdk-nodejs.png?branch=master)](https://travis-ci.org/blocktrail/blocktrail-sdk-nodejs)
[![Sauce Test Status](https://saucelabs.com/buildstatus/team_blocktrail)](https://saucelabs.com/u/team_blocktrail)


[![Sauce Test Status](https://saucelabs.com/browser-matrix/team_blocktrail.svg)](https://saucelabs.com/u/team_blocktrail)

The Blocktrail SDK is tested against;  
 - NodeJS:
   - 0.11
   - 0.12
   - 5.11
   - 6.3.0
   - 7.1.0
 - Browser:
   - Google Chrome 48 / latest
   - Firefox 49 / latest
   - Safari 10.0 / latest
   - Edge 14.14393
   - IE 11.103
   - Android 4.4
   - Android 5.0
   - iPhone OS X 10.10


Upgrading from v1.x to v2.0.0
-----------------------------
**IMPORTANT** `v2.0.0` has a few BC breaks, please check [docs/CHANGELOG.md](docs/CHANGELOG.md)!!

IMPORTANT! FLOATS ARE EVIL!!
----------------------------
As is best practice with financial data, The API returns all values as an integer, the Bitcoin value in Satoshi's.
**In Javascript even more than in other languages it's really easy to make mistakes when converting from float to integer etc!**

The BlockTrail SDK has some easy to use functions to do this for you, we recommend using these
and we also **strongly** recommend doing all Bitcoin calculation and storing of data in integers
and only convert to/from Bitcoin float values for displaying it to the user.

```javascript
var blocktrail = require('blocktrail-sdk');

console.log("123456789 Satoshi to BTC: ", blocktrail.toBTC(123456789));
console.log("1.23456789 BTC to Satoshi: ", blocktrail.toSatoshi(1.23456789));
```

A bit more about this can be found [in our documentation](https://www.blocktrail.com/api/docs/lang/nodejs#api_coin_format).

Installation
------------
You can install the package through NPM (https://www.npmjs.org/package/blocktrail-sdk).
```
npm install blocktrail-sdk
```

Usage
-----
Please visit our official documentation at https://www.blocktrail.com/api/docs/lang/nodejs for the usage.

Promises vs Callbacks
---------------------
Personally we prefer good old callbacks over promises,  
but to make everyone happy the SDK functions accept a callback argument and return a (Q)promise object, so you can use whatever you prefer!

Support and Feedback
--------------------
Be sure to visit the BlockTrail API official [documentation website](https://www.blocktrail.com/api/docs/lang/nodejs)
for additional information about our API.

If you find a bug, please submit the issue in Github directly.
[BlockTrail-NodeJS-SDK Issues](https://github.com/blocktrail/blocktrail-sdk-nodejs/issues)

As always, if you need additional assistance, drop us a note at
[support@blocktrail.com](mailto:support@blocktrail.com).

Unit Tests
----------
Unit Tests are created with Mocha and can be ran with `npm test` (or `mocha`)

We also run jshint and jscs, these are automatically ran by [travis-ci](https://travis-ci.org/blocktrail/blocktrail-sdk-nodejs) for every commit and pull request.
```
jshint main.js lib/ test/
```
```
jscs main.js lib/ test/
```

Browserify
----------
The BlockTrail NodeJS SDK can be browserified to use it in the browser, which we use ourselves for wallet actions from our webapp.  
If you want to test or develop on the SDK in the browser you can use `grunt build` (and `grunt watch`) to build the browserify version.

You need to pull the git submodules before you can build the browserify version:  
`git submodule update --init --recursive`

Files in `./build` are:

 - `blocktrail-sdk.js` (and `blocktrail-sdk.min.js`) the blocktrailSDK browserified  
 - `jsPDF.js` (and `jsPDF.min.js`) dependancy for generating the backup PDF  
 - `blocktrail-sdk-full.js` (and `blocktrail-sdk-full.min.js`) the blocktrailSDK browserified + jsPDF bundled  

If you use these browserified versions of our SDK it will be accessible as `window.blocktrailSDK` or plain `blocktrailSDK`.

Known Supported and Tested Browsers:
 - Android 4.3+ (Use Crosswalk for lower versions if neccesary)
 - iOS6+
 - IE9+

For the following any modern version will work just fine:
 - Chrome
 - FireFox
 - Safari

Uglify
------
If you're planning to uglify/minify the javascript yourself, make sure to exclude the following variable names from being mangled:  
`['Buffer', 'BitInteger', 'Point', 'Script', 'ECPubKey', 'ECKey']`

License
-------
The BlockTrail NodeJS SDK is released under the terms of the MIT license. See LICENCE.md for more information or see http://opensource.org/licenses/MIT.



docs/CHANGELOG.md
BlockTrail NodeJS SDK Changelog
===============================

v3.0.0
------
 - New [Default] Wallet Version 3  
   Better encryption scheme / key derivation.

v2.0.1
------
 - Add support for `OP_RETURN` in `wallet.pay` (see `examples/opreturn_payment_api_usage.js`)

v2.0.0
------
 - New [Default] Wallet Version 2
 - `BackupGenerator` now supports `extra` to be printed on document for extra notes
 - No longer support `BackupGenerator::generateImage`
 - 2FA support
 - Floating Fees

### Upgrade / BC breaks
 - `createNewWallet` now returns `spread(wallet, backupInfo)` to be able to support both v1 and v2
 - `new BackupGenerator()` now takes `identifier, backupInfo, extra`
 - No longer support `BackupGenerator::generateImage`
 - Defaults to using `Wallet.FEE_STRATEGY_OPTIMAL` instead of `Wallet.FEE_STRATEGY_BASE_FEE`!

### New [Default] Wallet Version 2
Instead of using `BIP39`, wallet seeds will now be stored encrypted - to allow for password changes

Wallet Creation:  
```
primarySeed = random()
secret = random()
primaryMnemonic = BIP39.entropyToMnemonic(AES.encrypt(primarySeed, secret))
secretMnemonic = BIP39.entropyToMnemonic(AES.encrypt(secret, password))
```

Wallet Init:  
```
secret = BIP39.entropyToMnemonic(AES.decrypt(secretMnemonic, password))
primarySeed = BIP39.entropyToMnemonic(AES.decrypt(primaryMnemonic, secret))
```

See `docs/KEYS.md` for more info
   
Old Wallets that are v1 will remain so and will continue working.

Wallets created through the Blocktrail website will have an option to choose which version to create (default v2), 
as grace period the first few days after the SDK release this will default to v1 still.

### Floating Fees
To deal with stresstests etc. there's now a `feePerKB` method to get the optimal fee and the `wallet.pay` has a `feeStrategy` argument.  
When `feeStrategy` is `Wallet.FEE_STRATEGY_OPTIMAL` (default) it will use the (by the API calculated) optimal fee per KB.  
When `feeStrategy` is `Wallet.FEE_STRATEGY_BASE_FEE` it will use the BASE_FEE of 0.0001 BTC per KB and use the old way of rounding the transaction size to 1 KB.

Optimal fee is calculated by taking all transactions from the last 30 blocks and calculating what the lowest possible fee is 
that still gives more than 80% chance to end up in the next block.

### 2FA Support
When enabled (currently still impossible to enable) you'll be required to add a `twoFactorToken` to the `wallet.pay`.  
We'll do another patch soon to add ways to setup 2FA and add it to the GUI on the site!

v1.3.12
-------
 - add batch support for fetching multiple transactions.
 - use .notify on the pay promise for progress.
 - allow for bypassing of local derivation of new address (used to verify API response)


Gruntfile.js
module.exports = function (grunt) {

    var browsers = [{
        // latest chrome
        browserName: 'googlechrome',
        platform: 'Windows 10',
        version: 'latest'
    }, {
        // latest chrome as of writing
        browserName: 'googlechrome',
        platform: 'Windows 10',
        version: '54.0'
    }, {
        // latest firefox
        browserName: 'firefox',
        platform: 'Windows 10',
        version: 'latest'
    }, {
        // latest firefox as of writing
        browserName: 'firefox',
        platform: 'Windows 10',
        version: '49.0'
    }, {
        // latest Safari
        browserName: 'Safari',
        version: 'latest'
    }, {
        // latest Safari as of writing
        browserName: 'Safari',
        version: '10.0'
    }, {
        // latest edge
        browserName: 'MicrosoftEdge',
        platform: 'Windows 10',
        version: 'latest'
    }, {
        // latest edge as of writing
        browserName: 'MicrosoftEdge',
        platform: 'Windows 10',
        version: '14.14393'
    }, {
        // latest IE (EOL)
        browserName: 'internet explorer',
        platform: 'Windows 10',
        version: '11.103'
    }, {
        // android 5.0
        browserName: 'android',
        platform: 'Linux',
        version: '5.0'
    }, {
        // android 4.4
        browserName: 'android',
        platform: 'Linux',
        version: '4.4'
    }, {
        // iphone iOS 9.2
        browserName: 'iphone',
        platform: 'OS X 10.10',
        version: '9.2'
    }, {
        // on FF < 48 there's no crypto.getRandomValues in webworkers
        browserName: 'firefox',
        platform: 'Windows 10',
        version: '47'
    }];

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        connect: {
            server: {
                options: {
                    base: '',
                    port: 9999
                }
            }
        },

        'saucelabs-mocha': {
            all: {
                options: {
                    // username: 'saucelabs-user-name', // if not provided it'll default to ENV SAUCE_USERNAME (if applicable)
                    // key: 'saucelabs-key', // if not provided it'll default to ENV SAUCE_ACCESS_KEY (if applicable)
                    urls: [
                        'http://127.0.0.1:9999/test/run-tests.html'
                    ],
                    browsers: browsers,
                    build: process.env.TRAVIS_JOB_ID || ('99' + ((new Date).getTime() / 1000).toFixed(0) + (Math.random() * 1000).toFixed(0)),
                    testname: 'mocha tests',
                    throttled: 2,
                    statusCheckAttempts: 180, // statusCheckAttempts * pollInterval = total time
                    pollInterval: 4000,
                    sauceConfig: {
                        'video-upload-on-pass': true
                    }
                }
            }
        },

        template: {
            runtests: {
                options: {
                    data: {
                        process: {
                            env: process.env
                        }
                    }
                },
                files: {
                    'test/run-tests.html': ['test/run-tests.tpl.html']
                }
            }
        },

        exec: {
            asmcryptobuild: 'cd ./vendor/asmcrypto.js; npm install; grunt --with pbkdf2-hmac-sha512'
        },

        /*
         * Javascript concatenation
         */
        concat : {
            jsPDF: {
                src : [
                    'vendor/jsPDF/jspdf.js',
                    'vendor/jsPDF/jspdf.plugin.split_text_to_size.js',
                    'vendor/jsPDF/jspdf.plugin.addimage.js',
                    'vendor/jsPDF/libs/FileSaver.js/FileSaver.js',
                    'vendor/jsPDF/jspdf.plugin.png_support.js',
                    'vendor/jsPDF/libs/png_support/zlib.js',
                    'vendor/jsPDF/libs/png_support/png.js'
                ],
                dest : 'build/jsPDF.js'
            },
            sdkfull: {
                src : [
                    '<%= concat.jsPDF.dest %>',
                    '<%= browserify.sdk.dest %>'
                ],
                dest : 'build/blocktrail-sdk-full.js'
            }
        },

        /*
         * Javascript uglifying
         */
        uglify : {
            options: {
                mangle: {
                    except: ['Buffer', 'BitInteger', 'Point', 'Script', 'ECPubKey', 'ECKey']
                }
            },
            dist : {
                files : {
                    'build/jsPDF.min.js'                : ['<%= concat.jsPDF.dest %>'],
                    'build/blocktrail-sdk.min.js'       : ['<%= browserify.sdk.dest %>'],
                    'build/blocktrail-sdk-full.min.js'  : ['<%= concat.sdkfull.dest %>']
                }
            }
        },

        /*
         *
         */
        browserify: {
            sdk: {
                options : {
                    browserifyOptions : {
                        standalone: 'blocktrailSDK'
                    },
                    transform : ['brfs']
                },
                src: 'main.js',
                dest: 'build/blocktrail-sdk.js'
            },
            test: {
                options : {
                    browserifyOptions : {
                        standalone: 'blocktrailTEST'
                    },
                    transform : ['brfs']
                },
                src: 'test.js',
                dest: 'build/test.js'
            }
        },

        /*
         * Watch
         */
        watch : {
            options : {},
            gruntfile : {
                files : ['Gruntfile.js'],
                tasks : ['default']
            },
            browserify : {
                files : ['main.js', 'test.js', 'test/*', 'test/**/*', 'lib/*', 'lib/**/*'],
                tasks : ['browserify', 'concat', 'template']
            },
            deps : {
                files : ['vendor/**/*.js'],
                tasks : ['concat']
            }
        }
    });

    grunt.loadNpmTasks('grunt-browserify');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-connect');
    grunt.loadNpmTasks('grunt-saucelabs');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-contrib-concat');
    grunt.loadNpmTasks('grunt-notify');
    grunt.loadNpmTasks('grunt-template');
    grunt.loadNpmTasks('grunt-exec');

    grunt.registerTask('dev', ['browserify', 'concat:sdkfull']);
    grunt.registerTask('asmcrypto', ['exec:asmcryptobuild']);
    grunt.registerTask('build', ['asmcrypto', 'browserify', 'concat', 'uglify']);
    grunt.registerTask('test-browser', ['template', 'connect', 'saucelabs-mocha']);
    grunt.registerTask('default', ['build']);
};

